File: agentsdk/python/agentsdk/events.py
import json
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any, Dict, Optional


class EventType(str, Enum):
    SESSION_START = "session.start"
    SESSION_END = "session.end"
    AGENT_START = "agent.start"
    AGENT_END = "agent.end"
    TOOL_START = "tool.start"
    TOOL_END = "tool.end"
    LLM_START = "llm.start"
    LLM_END = "llm.end"
    THINKING = "agent.thinking"
    CUSTOM = "custom"


@dataclass
class Event:
    type: EventType
    timestamp: datetime
    session_id: str = ""
    sequence: int = 0
    data: Dict[str, Any] = field(default_factory=dict)
    
    def to_json(self) -> str:
        return json.dumps({
            "type": self.type.value,
            "timestamp": self.timestamp.isoformat(),
            "session_id": self.session_id,
            "sequence": self.sequence,
            "data": self.data,
        })
    
    @classmethod
    def from_json(cls, data: str) -> "Event":
        obj = json.loads(data)
        return cls(
            type=EventType(obj["type"]),
            timestamp=datetime.fromisoformat(obj["timestamp"]),
            session_id=obj.get("session_id", ""),
            sequence=obj.get("sequence", 0),
            data=obj.get("data", {}),
        )

File: agentsdk/go/events.go
package agentsdk

import (
	"encoding/json"
	"time"
)

type EventType string

const (
	EventSessionStart EventType = "session.start"
	EventSessionEnd   EventType = "session.end"
	EventAgentStart   EventType = "agent.start"
	EventAgentEnd     EventType = "agent.end"
	EventToolStart    EventType = "tool.start"
	EventToolEnd      EventType = "tool.end"
	EventLLMStart     EventType = "llm.start"
	EventLLMEnd       EventType = "llm.end"
	EventThinking     EventType = "agent.thinking"
	EventCustom       EventType = "custom"
)

type Event struct {
	Type      EventType              `json:"type"`
	Timestamp time.Time              `json:"timestamp"`
	SessionID string                 `json:"session_id"`
	Sequence  int64                  `json:"sequence"`
	Data      map[string]interface{} `json:"data,omitempty"`
}

func (e *Event) ToJSON() ([]byte, error) {
	return json.Marshal(e)
}

func EventFromJSON(data []byte) (*Event, error) {
	var e Event
	if err := json.Unmarshal(data, &e); err != nil {
		return nil, err
	}
	return &e, nil
}

